protocol_driver_grpc_async_callback.cc

```diff
@@ -316,4 +323,167 @@ void ProtocolDriverGrpcAsyncCallback::ShutdownServer() {
   }
 }
 
+using grpc::Status;
+
+class CallData {
+ public:
+  CallData(Traffic::AsyncService* service, grpc::ServerCompletionQueue* cq,
+           std::function<std::function<void()>(ServerRpcState* state)>* handler,
+           DistbenchThreadpool* thread_pool)
+      : service_(service), cq_(cq),
+        handler_(handler), responder_(&ctx_),
+        status_(CREATE) {
+    CHECK(thread_pool);
+    thread_pool_ = thread_pool;
+    ProcessRpcFsm();
+  }
+
+  void Unref() {
+    if (--refcnt_ == 0) {
+      delete this;
+    }
+  }
+
+  void ProcessGenericRpc() {
+    rpc_state_.have_dedicated_thread = false;
+    rpc_state_.request = &request_;
+    rpc_state_.SetSendResponseFunction(
+        [&]() {
+          response_ = std::move(rpc_state_.response);
+          responder_.Finish(response_, Status::OK, this);
+        });
+    rpc_state_.SetFreeStateFunction([=]() { Unref();});
+    if(*handler_) {
+      auto f = (*handler_)(&rpc_state_);
+      if (f) {
+        thread_pool_->AddWork(f);
+      }
+    }
+  }
+
+  void ProcessRpcFsm() {
+    if (status_ == CREATE) {
+      status_ = PROCESS;
+      service_->RequestGenericRpc(&ctx_, &request_, &responder_, cq_, cq_, this);
+    } else if (status_ == PROCESS) {
+      status_ = FINISH;
+      new CallData(service_, cq_, handler_, thread_pool_);
+      ProcessGenericRpc();
+    } else {
+      GPR_ASSERT(status_ == FINISH);
+      Unref();
+    }
+  }
+
+  private:
+  GenericRequest request_;
+  GenericResponse response_;
+  Traffic::AsyncService* service_;
+  grpc::ServerCompletionQueue* cq_;
+  std::function<std::function<void()>(ServerRpcState* state)>* handler_;
+  grpc::ServerAsyncResponseWriter<GenericResponse> responder_;
+  grpc::ServerContext ctx_;
+  enum CallStatus { CREATE, PROCESS, FINISH };
+  CallStatus status_;
+  DistbenchThreadpool* thread_pool_;
+  ServerRpcState rpc_state_;
+  std::atomic<int> refcnt_ = 2;
+};
+
+// Server =====================================================================
+namespace {
+class TrafficAsyncService : public Traffic::AsyncService {
```

Comment:
There is both TrafficAsyncService and TrafficServiceAsync; we probably should rename one of them to avoid confusion. Maybe the other one ? TrafficServiceAsync => TrafficCallbackService ?